### PR TITLE
[OpenVINO] Support Qwen3-next

### PR DESCRIPTION
**Pull request by rkazants**
# What does this PR do?

Example of conversion cmd-line for `Qwen/Qwen3-Next-80B-A3B-Instruct`:

```sh
optimum-cli export openvino -m Qwen/Qwen3-Next-80B-A3B-Instruct Qwen3-Next-80B-A3B-Instruct
```

Example of inference for `Qwen/Qwen3-Next-80B-A3B-Instruct` using OpenVINO backend:

```python
from transformers import AutoTokenizer
from optimum.intel.openvino import OVModelForCausalLM

model_path = "./Qwen3-Next-80B-A3B-Instruct"

tokenizer = AutoTokenizer.from_pretrained(model_path)
model = OVModelForCausalLM.from_pretrained(model_path, trust_remote_code=True)

# change input text as desired
input_text = "The capital of France is"
# tokenize the text
input_tokens = tokenizer(input_text, return_tensors="pt")
# generate output tokens
output = model.generate(**input_tokens, max_length=10)
# decode output tokens into text
output = tokenizer.batch_decode(output)
print(output[0])
```

## Before submitting
- [N/A] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/exporters/openvino/model_patcher.py [hunk 1]
```diff
@@ -7409,3 +7409,723 @@ def unpatch_sparse_moe(sparse_moe_layer):
...
+            mixed_qkv_dec,
+            conv_state_dec,
+            self.conv1d.weight.squeeze(1),
+            self.conv1d.bias,
+            self.activation,
+        )
+
+        conv_state_prefill = F.pad(mixed_qkv, (self.conv_kernel_size - mixed_qkv.shape[-1], 0))
+        conv_state = conv_state_dec * use_precomputed_states + conv_state_prefill * (1.0 - use_precomputed_states)
+        mixed_qkv = mixed_qkv_dec * use_precomputed_states + mixed_qkv * (1.0 - use_precomputed_states)
```
Comment (zhangYiIntel) [inline optimum/exporters/openvino/model_patcher.py]: Prefill path misses the conv1d.
```suggestion
        mixed_qkv_prefill = F.silu(self.conv1d(mixed_qkv)[:, :, :seq_len])
        mixed_qkv = mixed_qkv_dec * use_precomputed_states + mixed_qkv_prefill * (1.0 - use_precomputed_states)
```

Comment (rkazants) [reply optimum/exporters/openvino/model_patcher.py]: Done

File: tests/openvino/utils_tests.py [hunk 2]
```diff
@@ -370,6 +371,7 @@
     "lfm2": {"model": 52},
     "hunyuan_v1_dense": {"model": 32},
     "qwen3_eagle3": {"model": 20},
+    "qwen3_next": {"model": 20},
```
Comment (rkazants) [inline tests/openvino/utils_tests.py]:
```suggestion
    "qwen3_next": {"model": 100},
```

File: optimum/exporters/openvino/model_patcher.py [hunk 3]
```diff
@@ -7899,3 +7899,512 @@ def forward(
...
+    router_logits = self.gate(hidden_states)
+
+    routing_weights = F.softmax(router_logits, dim=1, dtype=torch.float)
+    routing_weights, selected_experts = torch.topk(routing_weights, self.top_k, dim=-1)
+    if self.norm_topk_prob:
+        routing_weights /= routing_weights.sum(dim=-1, keepdim=True)
+    # we cast back to the input dtype
+    routing_weights = routing_weights.to(hidden_states.dtype)
+
+    new_routing_weights = torch.zeros(batch_size * sequence_length, num_experts, dtype=routing_weights.dtype)
```
[the remark this answers is not in the document]
Comment (rkazants) [reply optimum/exporters/openvino/model_patcher.py:8081]: not needed, it is always on CPU device

File: docs/source/openvino/models.mdx [hunk 4]
```diff
@@ -131,6 +131,7 @@ Here is the list of the supported architectures :
 - Qwen2VL
 - Qwen2.5VL
 - Qwen3VL
+- Qwen3-Next
```
[the remark this answers is not in the document]
Comment (rkazants) [reply docs/source/openvino/models.mdx:134]: Done

File: optimum/exporters/openvino/model_patcher.py [hunk 5]
```diff
@@ -7899,3 +7899,512 @@ def forward(
...
+        key = key.repeat_interleave(self.num_v_heads // self.num_k_heads, dim=2)
+
+    core_attn_out, last_recurrent_state = self.recurrent_gated_delta_rule(
+        self,
+        query,
+        key,
+        value,
+        g=g,
+        beta=beta,
+        initial_state=recurrent_state,
```
[the remark this answers is not in the document]
Comment (rkazants) [reply optimum/exporters/openvino/model_patcher.py:8044]: Fixed.

File: optimum/exporters/openvino/model_patcher.py [hunk 6]
```diff
@@ -7899,3 +7899,513 @@ def forward(
...
+    router_logits = self.gate(hidden_states)
+
+    routing_weights = F.softmax(router_logits, dim=1, dtype=torch.float)
+    routing_weights, selected_experts = torch.topk(routing_weights, self.top_k, dim=-1)
+    if self.norm_topk_prob:
+        routing_weights /= routing_weights.sum(dim=-1, keepdim=True)
+    # we cast back to the input dtype
+    routing_weights = routing_weights.to(hidden_states.dtype)
+
+    new_routing_weights = torch.zeros(batch_size * sequence_length, num_experts, dtype=routing_weights.dtype)
```
[the remark this answers is not in the document]
Comment (rkazants) [reply optimum/exporters/openvino/model_patcher.py:8081]: OpenVINO does not support cuda, not needed

File: optimum/exporters/openvino/convert.py [hunk 7]
```diff
@@ -435,11 +435,16 @@ def ts_patched_forward(*args, **kwargs):
...
+                module_extensions = getattr(patcher, "module_extensions", None)
+                if module_extensions is not None:
+                    ts_decoder_kwargs["module_extensions"] = module_extensions
                 ts_decoder = TorchScriptPythonDecoder(model, example_input=dummy_inputs, **ts_decoder_kwargs)
                 ov_model = convert_model(
                     ts_decoder,
                     example_input=dummy_inputs,
                     input=[(item.shape, item.type) for item in input_info],
+                    extension=conversion_extensions,
                 )
```
[the remark this answers is not in the document]
Comment (rkazants) [reply optimum/exporters/openvino/convert.py:448]: For torch dynamo export, the model is not yet supported and module extension mechanism is not needed

File: optimum/exporters/openvino/model_patcher.py [hunk 8]
```diff
@@ -7899,3 +7899,513 @@ def forward(
...
+
+        super().__exit__(exc_type, exc_value, traceback)
+        setattr(self._model, self.orig_forward_name, self.model_orig_forward)
+        for idx, decoder_layer in enumerate(self._model.model.layers):
+            layer_type = self._model.model.config.layer_types[idx]
+            if layer_type == "linear_attention":
+                linear_attn_layer = decoder_layer.linear_attn
+                linear_attn_layer.forward = linear_attn_layer._orig_forward
+                linear_attn_layer.chunk_gated_delta_rule = linear_attn_layer._orig_chunk_gated_delta_rule
+            if isinstance(decoder_layer.mlp, Qwen3NextSparseMoeBlock):
```
[the remark this answers is not in the document]
Comment (rkazants) [reply optimum/exporters/openvino/model_patcher.py:8318]: Removed.

File: optimum/exporters/openvino/model_configs.py [hunk 9]
```diff
@@ -5314,3 +5315,142 @@ class HunyuanV1DenseOpenVINOConfig(LlamaOpenVINOConfig):
...
+            cache_name_prefix = "cache_params.past"
+        else:
+            decoder_sequence_name = "past_sequence_length + sequence_length"
+            cache_name_prefix = "cache_params.present"
+
+        self.num_full_attn_layers = self._normalized_config.layer_types.count("full_attention")
+        self.num_linear_attn_layers = self._normalized_config.layer_types.count("linear_attention")
+
+        for i in range(self.num_linear_attn_layers):
+            # [batch_size, conv_kernel_size - 1, d_model]
```
Comment (rkazants) [inline optimum/exporters/openvino/model_configs.py]:
```suggestion
```

File: optimum/exporters/openvino/model_configs.py [hunk 10]
```diff
@@ -5314,3 +5315,141 @@ class HunyuanV1DenseOpenVINOConfig(LlamaOpenVINOConfig):
...
+        else:
+            decoder_sequence_name = "past_sequence_length + sequence_length"
+            cache_name_prefix = "cache_params.present"
+
+        self.num_full_attn_layers = self._normalized_config.layer_types.count("full_attention")
+        self.num_linear_attn_layers = self._normalized_config.layer_types.count("linear_attention")
+
+        for i in range(self.num_linear_attn_layers):
+            inputs_or_outputs[f"{cache_name_prefix}.conv.{i}"] = {0: "batch_size"}
+            # [batch_size, d_state, d_model]
```
Comment (rkazants) [inline optimum/exporters/openvino/model_configs.py]:
```suggestion
```

File: optimum/exporters/openvino/model_configs.py [hunk 11]
```diff
@@ -5314,3 +5315,140 @@ class HunyuanV1DenseOpenVINOConfig(LlamaOpenVINOConfig):
...
+        self.head_v_dim = config.linear_value_head_dim
+        self.num_v_heads = config.linear_num_value_heads
+        self.num_k_heads = config.linear_num_key_heads
+        self.num_key_value_heads = config.num_key_value_heads
+
+    def generate(self, input_name: str, framework: str = "pt", int_dtype: str = "int64", float_dtype: str = "fp32"):
+        cache_params = []
+
+        for idx in range(self.num_linear_attn_layers):
+            # (batch_size, d_inner, d_conv)
```
Comment (rkazants) [inline optimum/exporters/openvino/model_configs.py]:
```suggestion
```

File: docs/source/openvino/models.mdx (hunk 4 above)
Comment (rkazants) [inline docs/source/openvino/models.mdx:134]:
```suggestion
- Qwen3-Next 
```

File: docs/source/openvino/models.mdx [hunk 12]
```diff
@@ -131,6 +131,7 @@ Here is the list of the supported architectures :
...
```
Comment (rkazants) [inline docs/source/openvino/models.mdx]:
```suggestion
- Qwen3-Next
```

File: optimum/exporters/openvino/model_patcher.py [hunk 13]
```diff
@@ -7899,3 +7899,513 @@ def forward(
...
+    cache_params=None,
+    cache_position: Optional[torch.LongTensor] = None,
+    attention_mask: Optional[torch.Tensor] = None,
+):
+    def apply_mask_to_padding_states(hidden_states, attention_mask):
+        """
+        Tunes out the hidden states for padding tokens, see https://github.com/state-spaces/mamba/issues/66
+        """
+        # NOTE: attention mask is a 2D boolean tensor
+        if attention_mask is not None and attention_mask.shape[1] > 1 and attention_mask.shape[0] > 1:
```
[the remark this answers is not in the document]
Comment (rkazants) [reply optimum/exporters/openvino/model_patcher.py:7982]: Not needed that is because attention mask is guaranteed to be 2D

File: optimum/exporters/openvino/model_patcher.py [hunk 14]
```diff
@@ -7899,3 +7899,513 @@ def forward(
...
+        # the module to produce only one output.
+        output_cell = torch.cat([core_attn_out.flatten(), last_recurrent_state.flatten()], dim=0)
+        return output_cell
+
+
+# Conversion rule for the `RecurrentAttentionCellOp` operation in a Torch graph.
+# The `RecurrentAttentionCellOp` appears in the Torch graph as a result of replacing
+# the `torch.nn.Module` block `RecurrentAttentionCell` via a registered
+# `ModuleExtension` for `RecurrentAttentionCell` in the OpenVINO PyTorch frontend.
+def convert_recurrent_attention_cell(context):
```
Comment (IlyasMoutawwakil) [inline optimum/exporters/openvino/model_patcher.py]: why are we doing direct ov implementation here ? instead of the usual traceable replacement

Comment (rkazants) [reply optimum/exporters/openvino/model_patcher.py]: That is because there is no vectorized form for GatedDeltaNet implementation for correct torch.jit tracing. That is why I use module extension mechanism.

Comment (IlyasMoutawwakil) [reply optimum/exporters/openvino/model_patcher.py]: can we please put it in a separate file

Comment (IlyasMoutawwakil) [reply optimum/exporters/openvino/model_patcher.py]: _ov_ops.py or something like that

File: optimum/exporters/openvino/model_patcher.py [hunk 15]
```diff
@@ -7899,3 +7899,513 @@ def forward(
             hidden_states=outputs.hidden_states,
             d2t=d2t_out,
         )
+
+
+def patched_recurrent_gated_delta_rule(
```
Comment (popovaan) [inline optimum/exporters/openvino/model_patcher.py:7914]: Could you please add a link to the original method?

File: optimum/exporters/openvino/model_patcher.py (hunk 14 above)
Comment (rkazants) [reply optimum/exporters/openvino/model_patcher.py]: @IlyasMoutawwakil, sure, I will do it.

File: optimum/exporters/openvino/model_patcher.py (hunk 15 above)
Comment (rkazants) [reply optimum/exporters/openvino/model_patcher.py:7914]: done and added a comment.

File: optimum/exporters/openvino/model_patcher.py (hunk 14 above)
Comment (rkazants) [reply optimum/exporters/openvino/model_patcher.py]: Done

File: optimum/exporters/openvino/model_patcher.py [hunk 16]
```diff
@@ -7899,3 +7901,421 @@ def forward(
...
+    def __exit__(self, exc_type, exc_value, traceback):
+        from transformers.models.qwen3_next.modeling_qwen3_next import Qwen3NextSparseMoeBlock
+
+        super().__exit__(exc_type, exc_value, traceback)
+        setattr(self._model, self.orig_forward_name, self.model_orig_forward)
+        for idx, decoder_layer in enumerate(self._model.model.layers):
+            layer_type = self._model.model.config.layer_types[idx]
+            if layer_type == "linear_attention":
+                linear_attn_layer = decoder_layer.linear_attn
+                linear_attn_layer.forward = linear_attn_layer._orig_forward
```
Comment (echarlaix) [inline optimum/exporters/openvino/model_patcher.py]: shouldn't we unpatch `recurrent_gated_delta_rule` as well ?
